README.rst
VTK Visualizer Control
======================

Easy visualization of point clouds and geometric primitives based on VTK

See http://www.skotheim.net/wiki/Python_VTK for more information

changelog.txt
0.9.6
=====
* Should now work with Python 3.6 (thanks to Olivier Roulet-Dubonnet)
* Optimizations that make it much faster to visualize point clouds with color

0.9.5
=====
* Changed to use PyQt5 instead of PyQt4

0.9.4
=====
* More fixes for VTK 6

0.9.3
=====
* Fixed incompatibility for VTK 6 in plotxyzrgb function

0.9.2
=====
* Fixed incompatibility for VTK 6 in plotxyz function

0.9.1
=====

* Removed dependency of python_qt_binding
* Removed dependency of vtk (since it is not installable from PyPI)
* Fixed bug: Closing a window that had already gone out of scope

0.9.0
=====
Initial version

test/test_vtk_visualizer.py
# -*- coding: utf-8 -*-
import unittest
import numpy as np
import vtk
import os
import os.path
from vtk_visualizer import *

src_dir = os.path.dirname(os.path.abspath(str(__file__)))
 
def EstimateNormals(pc): 
    # Differentiate in x direction (dx)
    dxx = np.diff(pc[:,:,0],axis=0)[:,1:]
    dxy = np.diff(pc[:,:,1],axis=0)[:,1:]
    dxz = np.diff(pc[:,:,2],axis=0)[:,1:]
    
    # Differentiate in y direction (dy)
    dyx = np.diff(pc[:,:,0],axis=1)[1:,:]
    dyy = np.diff(pc[:,:,1],axis=1)[1:,:]
    dyz = np.diff(pc[:,:,2],axis=1)[1:,:]
    
    # Calculate normals by cross product of dx*dy
    normals = np.zeros((pc.shape[0],pc.shape[1],3))    
    normals[1:,1:,0] = - dxy * dyz + dxz*dyy;
    normals[1:,1:,1] = - dxz * dyx + dxx*dyz;
    normals[1:,1:,2] = - dxx * dyy + dxy*dyx;
    
    # Normalize length of normals (add a small number to avoid division by 0)
    L = np.sqrt(normals[:,:,0]**2+normals[:,:,1]**2+normals[:,:,2]**2) + 1e-12
    normals[:,:,0] /= L
    normals[:,:,1] /= L
    normals[:,:,2] /= L
        
    return normals

class TestVTKVisualizer(unittest.TestCase):    
    def setUp(self):
        self.vtkControl = get_vtk_control()
        
        h=256; w=256
        
        [y,x] = np.mgrid[0:h,0:w].astype(np.float64)
        z = 10*np.cos(0.1*np.sqrt((x-50)**2+(y-50)**2))
        
        pc = np.zeros((h,w,3))
        pc[:,:,0] = x
        pc[:,:,1] = y
        pc[:,:,2] = z
            
        normals = EstimateNormals(pc)
        
        self.pts = pc.reshape(-1,3)
        self.pts_n = np.dstack([pc,normals])
        self.pts_n = self.pts_n.reshape(-1,6)
                
    def test_pointcloud(self):
        self.vtkControl.AddPointCloudActor(self.pts)

    def test_pointcloud_color(self):
        pts = np.random.rand(1000,6)     
        pts[:,:3] *= 10.0        
        pts[:,3:6] *= 255.0
        self.vtkControl.AddColoredPointCloudActor(pts)
        
    def test_actor(self):
        cone = vtk.vtkConeSource()      
        mapper = vtk.vtkPolyDataMapper()
        mapper.SetInputConnection(cone.GetOutputPort())        
        actor = vtk.vtkActor()
        actor.SetMapper(mapper)
        self.vtkControl.AddActor(actor)
        
    def test_box(self):
        self.vtkControl.AddBox([-10,-5,-10,-5,-10,0])
        
    def test_axes(self):
        self.vtkControl.AddAxesActor(10.0)
    
    def test_cylinder(self):
        self.vtkControl.AddCylinder((10,0,0),2.0,10.0)        
        
    def test_sphere(self):        
        self.vtkControl.AddSphere((-10,-10,10), 5.0)        
        
    def test_plane(self):
        self.vtkControl.AddPlane() 
        nID = self.vtkControl.GetLastActorID()
        self.vtkControl.SetActorColor(nID,(0,1,0))
        self.vtkControl.SetActorOpacity(nID, 0.5)
        self.vtkControl.SetActorScale(nID, (10,10,0))
        
    def test_hedgehog(self):
        self.vtkControl.AddHedgeHogActor(self.pts_n, 5.0)

    def test_hedgehog_scalars(self):
        pts_n = np.zeros((1000,7))        
        pts_n[:,:2] = np.random.rand(1000,2) * 10.0
        pts_n[:,5] = 1
        self.vtkControl.AddHedgeHogActorWithScalars(pts_n, 5.0)
        
    def test_shaded_points(self):
        self.vtkControl.AddShadedPointsActor(self.pts_n)
        
    def test_normals(self):
        pts_n = np.zeros((1000,6))
        pts_n[:,:2] = np.random.rand(1000,2) * 10.0
        pts_n[:,5] = 1
        self.vtkControl.AddNormalsActor(pts_n, 5.0)
        
    def test_ply(self):
        ply_file = os.path.join(src_dir,'teapot.ply')
        self.vtkControl.AddPLYActor(ply_file)

    def test_stl(self):
        stl_file = os.path.join(src_dir,'teapot.stl')
        self.vtkControl.AddSTLActor(stl_file)
                
if __name__ == '__main__':
       
    suite = unittest.TestLoader().loadTestsFromTestCase(TestVTKVisualizer)
    unittest.TextTestRunner(verbosity=2).run(suite)    
    unittest.main()


vtk_visualizer/pointobject.py
# -*- coding: utf-8 -*-
"""
VTK visualization object

@author: Øystein Skotheim, SINTEF ICT <oystein.skotheim@sintef.no>
"""

import vtk
from vtk.util import numpy_support
import numpy as np

class VTKObject:
    """VTK visualization object
    Class that sets up the necessary VTK pipeline for displaying
    various objects (point clouds, meshes, geometric primitives)"""

    def __init__(self):    
        self.verts = None
        self.cells = None
        self.scalars = None
        self.normals = None
        self.pd = None
        self.LUT = None
        self.mapper = None
        self.actor = None
    
    def CreateFromSTL(self,filename):
        "Create a visualization object from a given STL file"
                        
        if not filename.lower().endswith('.stl'):
            raise Exception('Not an STL file')
            
        reader = vtk.vtkSTLReader()
        reader.SetFileName(filename)
        reader.Update()
        
        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(reader.GetOutput())
        
        self.SetupPipelineMesh()
        
    def CreateFromPLY(self,filename):
        "Create a visualization object from a given PLY file"
                        
        if not filename.lower().endswith('.ply'):
            raise Exception('Not a PLY file')
            
        reader = vtk.vtkPLYReader()
        reader.SetFileName(filename)
        reader.Update()
        
        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(reader.GetOutput())
        
        self.SetupPipelineMesh()
        
    def CreateFromActor(self,actor):
        "Create a visualization object from a given vtkActor instance"
        if not isinstance(actor,vtk.vtkObject):
            raise Exception('Argument is not a VTK Object')        
        self.actor = actor
                    
    def CreateFromPolyData(self,pd):
        "Create a visualization object from a given vtkPolyData instance"        
        self.pd = pd
        self.SetupPipelineMesh()
        
    def CreateFromArray(self,pc):
        """Create a point cloud visualization object from a given NumPy array
        
        The NumPy array should have dimension Nxd where d >= 3
        
        If d>3, the points will be colored according to the last column
        in the supplied array (values should be between 0 and 1, where 
        0 is black and 1 is white)
        """
        
        nCoords = pc.shape[0]
        nElem = pc.shape[1]
        
        if nElem < 3:
            raise Exception('Number of elements must be greater than or equal to 3')

        self.verts = vtk.vtkPoints()
        self.cells = vtk.vtkCellArray()
        self.scalars = None
        self.pd = vtk.vtkPolyData()        
                    
        # Optimized version of creating the vertices array                               
        # - We need to be sure to keep the supplied NumPy array,
        #   because the numpy_support function creates a pointer into it
        # - We need to take a copy to be sure the array is contigous                    
        self.points_npy = pc[:,:3].copy()   
        self.verts.SetData(numpy_support.numpy_to_vtk(self.points_npy))
        
        # Optimized version of creating the cell ID array
        # - We need to be sure to keep the supplied NumPy array,
        #   because the numpy_support function creates a pointer into it    
        # - Note that the cell array looks like this: [1 vtx0 1 vtx1 1 vtx2 1 vtx3 ... ]
        #   because it consists of vertices with one primitive per cell
        self.cells_npy = np.vstack([np.ones(nCoords,dtype=np.int64),
                                  np.arange(nCoords,dtype=np.int64)]).T.flatten()

        self.cells.SetCells(nCoords,numpy_support.numpy_to_vtkIdTypeArray(self.cells_npy))

        self.pd.SetPoints(self.verts)
        self.pd.SetVerts(self.cells)
        
        # Optimized version of creating the scalars array
        # - We need to be sure to keep the NumPy array,
        #   because the numpy_support function creates a pointer into it        
        # - We need to take a copy to be sure the array is contigous                    
        if (nElem > 3):
            self.scalars_npy = pc[:,-1].copy()
            self.scalars = numpy_support.numpy_to_vtk(self.scalars_npy)
         
            # Color the scalar data in gray values                   
            self.LUT = vtk.vtkLookupTable()
            self.LUT.SetNumberOfColors(255)
            self.LUT.SetSaturationRange(0, 0)
            self.LUT.SetHueRange(0, 0)
            self.LUT.SetValueRange(0, 1)
            self.LUT.Build()
            self.scalars.SetLookupTable(self.LUT)
            self.pd.GetPointData().SetScalars(self.scalars)
            
        self.SetupPipelineCloud()
        
    def AddNormals(self,ndata):
        "Add surface normals (Nx3 NumPy array) to the current point cloud visualization object"
        
        nNormals = ndata.shape[0]
        nDim = ndata.shape[1]

        if (nDim != 3):
            raise Exception('Expected Nx3 array of surface normals')
        
        if (self.pd.GetNumberOfPoints() == 0):
            raise Exception('No points to add normals for')
            
        if (nNormals != self.pd.GetNumberOfPoints()):
            raise Exception('Supplied number of normals incompatible with number of points')

        # Optimized version of creating the scalars array
        # - We need to be sure to keep the NumPy array,
        #   because the numpy_support function creates a pointer into it        
        # - We need to take a copy to be sure the array is contigous                                
        self.normals_npy = ndata.copy()        
        self.normals = numpy_support.numpy_to_vtk(self.normals_npy)
                                    
        self.pd.GetPointData().SetNormals(self.normals)
        self.pd.Modified()
        
    def AddColors(self,cdata):
        """"Add colors (Nx3 NumPy array) to the current point cloud visualization object        
        NumPy array should be of type uint8 with R, G and B values between 0 and 255
        """
        
        nColors = cdata.shape[0]
        nDim = cdata.shape[1]
        
        if (self.pd.GetNumberOfPoints() == 0):
            raise Exception('No points to add color for')
            
        if (nColors != self.pd.GetNumberOfPoints()):
            raise Exception('Supplied number of colors incompatible with number of points')
            
        if (nDim != 3):
            raise Exception('Expected Nx3 array of colors')
            
        # Optimized version of creating the scalars array
        # - We need to be sure to keep the NumPy array,
        #   because the numpy_support function creates a pointer into it        
        # - We need to take a copy to be sure the array is contigous                                
        self.colors_npy = cdata.copy().astype(np.uint8)     
        self.colors = numpy_support.numpy_to_vtk(self.colors_npy)
                        
        self.pd.GetPointData().SetScalars(self.colors)
        self.pd.Modified()

    def SetupPipelineCloud(self):
        "Set up the VTK pipeline for visualizing a point cloud"
        
        self.mapper = vtk.vtkPolyDataMapper()
        self.mapper.SetInputData(self.pd)

        if (self.scalars != None):
            self.ScalarsOn()
            self.SetScalarRange(0, 1)

        self.actor = vtk.vtkActor()
        self.actor.SetMapper(self.mapper)
        self.actor.GetProperty().SetRepresentationToPoints()
        self.actor.GetProperty().SetColor(0.0,1.0,0.0)
        
    def SetupPipelineMesh(self):
        "Set up the VTK pipeline for visualizing a mesh"
        
        self.mapper = vtk.vtkPolyDataMapper()
        self.mapper.SetInputData(self.pd)
        
        if self.scalars != None:
            if self.pd.GetPointData() != None:
                if (self.pd.GetPointData().GetScalars != None) and (self.LUT != None):
                    self.pd.GetPointData().GetScalars().SetLookupTable(self.LUT)
                    
            self.mapper.ScalarVisibilityOn()
            self.mapper.SetColorModeToMapScalars()
            self.mapper.SetScalarModeToUsePointData()
            self.mapper.SetScalarRange(0,1)
            
        else:            
            self.mapper.ScalarVisibilityOff()

        self.actor = vtk.vtkActor()
        self.actor.SetMapper(self.mapper)
        self.actor.GetProperty().SetRepresentationToSurface()
        self.actor.GetProperty().SetInterpolationToGouraud()
        
    def SetupPipelineHedgeHog(self,scale=15.0):
        """Set up the VTK pipeline for visualizing points with surface normals"
        
        The surface normals are visualized as lines with the given scale"""
        
        hh = vtk.vtkHedgeHog()
        hh.SetInputData(self.pd)
        hh.SetVectorModeToUseNormal()
        hh.SetScaleFactor(scale)
        hh.Update()
        
        # Set up mappers and actors
        self.mapper = vtk.vtkPolyDataMapper()
        self.mapper.SetInputConnection(hh.GetOutputPort())
        self.mapper.Update()
        self.actor = vtk.vtkActor()
        self.actor.SetMapper(self.mapper)
        self.actor.GetProperty().SetColor(1,0,0) # Default color
                
    def ScalarsOn(self):
        "Enable coloring of the points based on scalar array"
        self.mapper.ScalarVisibilityOn()
        self.mapper.SetColorModeToMapScalars()
        self.mapper.SetScalarModeToUsePointData()
            
    def ScalarsOff(self):
        "Disable coloring of the points based on scalar array"
        self.mapper.ScalarVisibilityOff()
    
    def SetScalarRange(self, smin, smax):
        "Set the minimum and maximum values for the scalar array"
        self.mapper.SetScalarRange(smin, smax)
        
    def GetActor(self):
        "Returns the current actor"
        return self.actor
        
    # Primitives    
    def CreateSphere(self, origin, r):
        "Create a sphere with given origin (x,y,z) and radius r"
        sphere = vtk.vtkSphereSource()
        sphere.SetCenter(origin)
        sphere.SetRadius(r)
        sphere.SetPhiResolution(25)
        sphere.SetThetaResolution(25)
        sphere.Update()
        
        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(sphere.GetOutput())
        self.scalars = None
        self.SetupPipelineMesh()

    def CreateCylinder(self, origin, r, h):
        "Create a cylinder with given origin (x,y,z), radius r and height h"       
        cyl = vtk.vtkCylinderSource()
        cyl.SetCenter(origin)
        cyl.SetRadius(r)
        cyl.SetHeight(h)
        cyl.SetResolution(25)
        cyl.Update()
        
        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(cyl.GetOutput())
        self.scalars = None
        self.SetupPipelineMesh()
        
    def CreateAxes(self, length):
        "Create a coordinate axes system with a given length of the axes"               
        axesActor = vtk.vtkAxesActor()
        axesActor.AxisLabelsOff()
        axesActor.SetTotalLength(length, length, length)        
        self.actor = axesActor
        
    def CreatePlane(self, normal=None, origin=None):
        "Create a plane (optionally with a given normal vector and origin)"
        plane = vtk.vtkPlaneSource()
        plane.SetXResolution(10)
        plane.SetYResolution(10)
        if not normal is None: plane.SetNormal(normal)        
        if not origin is None: plane.SetCenter(origin)
        plane.Update()

        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(plane.GetOutput())
        self.scalars = None
        self.SetupPipelineMesh()
        
    def CreateBox(self, bounds):
        "Create a box with the given bounds [xmin,xmax,ymin,ymax,zmin,zmax]"
        box = vtk.vtkTessellatedBoxSource()
        box.SetBounds(bounds)
        box.Update()
        self.pd = box.GetOutput()
        self.scalars = None
        self.SetupPipelineMesh()
        
    def CreateLine(self, p1, p2):
        "Create a 3D line from p1=[x1,y1,z1] to p2=[x2,y2,z2]"
        line = vtk.vtkLineSource()
        line.SetPoint1(*p1)
        line.SetPoint2(*p2)
        line.Update()
        
        self.pd = vtk.vtkPolyData()
        self.pd.DeepCopy(line.GetOutput())
        self.scalars = None
        self.SetupPipelineMesh()
        
if __name__ == '__main__':
    
    # Create some random points
    pc = np.random.rand(10000,3)
    
    # Create some random normals
    normals = np.random.rand(10000,3)-0.5

    # Create some random colors
    colors = np.random.rand(10000,3)
    colors *= 255.0
    colors = colors.astype(np.uint8)
    
    obj = VTKObject()
    obj.CreateFromArray(pc)
    obj.AddColors(colors)
    obj.AddNormals(normals)
    obj.SetupPipelineHedgeHog(0.2)
    
    ren = vtk.vtkRenderer()
    ren.AddActor(obj.GetActor())
    
    renWin = vtk.vtkRenderWindow()
    renWin.AddRenderer(ren)
    
    iren = vtk.vtkRenderWindowInteractor()
    iren.SetRenderWindow(renWin)
    
    style = vtk.vtkInteractorStyleTrackballCamera()
    iren.SetInteractorStyle(style)
    
    iren.Initialize()
    iren.Start()

vtk_visualizer/renderwidget.py
# -*- coding: utf-8 -*-
"""
Render widget for VTK based on Qt

:Author: Øystein Skotheim, Zivid Labs <oystein.skotheim@zividlabs.com>
"""

import vtk

qt_api = 'pyqt'
try:
    from PyQt5.QtWidgets import *
except:
    qt_api = 'pyside'

if qt_api == 'pyside':
    try:
        from PySide import QtGui
    except:
        raise Exception('Need either PySide or PyQt4')
        
#from vtk.qt.QVTKRenderWindowInteractor import QVTKRenderWindowInteractor 
# Note: The QVTKRenderWindowInteractor included with VTK is buggy for PySide
# Use our modified version instead
from vtk_visualizer.QVTKRenderWindowInteractor import QVTKRenderWindowInteractor 

import sys

class RenderWidget:
    
    def __init__(self,renderer=None,parent=None):
        
        # Every QT app needs a QApplication
        self.app = QApplication.instance()
        if self.app is None:
            self.app = QApplication(sys.argv)
        
        # Create the widget
        if renderer is None:
            self.renderer = vtk.vtkRenderer()
        else:
            self.renderer = renderer
            
        self.widget = QVTKRenderWindowInteractor(parent)
        
        self.widget.Initialize
        self.widget.Start()

        # Set the interactor style 
        self.widget.SetInteractorStyle(vtk.vtkInteractorStyleTrackballCamera())
            
        # Get the render window from the widget 
        self.renderWindow =  self.widget.GetRenderWindow()
        self.renderWindow.AddRenderer(self.renderer)

        # show the widget
        self.widget.show()
        
    def exec_(self):
        self.app.exec_()
                
if __name__ == '__main__':    
    
    cone = vtk.vtkConeSource()
    cone.SetResolution(8)

    coneMapper = vtk.vtkPolyDataMapper()
    coneMapper.SetInputConnection(cone.GetOutputPort())

    coneActor = vtk.vtkActor()
    coneActor.SetMapper(coneMapper)

    ren = vtk.vtkRenderer()
    ren.AddActor(coneActor)
    
    w = RenderWidget(ren)
    w.exec_()


vtk_visualizer/visualizercontrol.py
# -*- coding: utf-8 -*-
"""
Easy visualization of point clouds and geometric primitives
Created on Fri Apr 08 12:01:56 2011

.. Author: Øystein Skotheim, SINTEF ICT <oystein.skotheim@sintef.no>
   Date:   Thu Sep 12 15:50:40 2013

"""

import vtk
from vtk_visualizer.renderwidget import RenderWidget
from vtk_visualizer.pointobject import *
                
class VTKVisualizerControl:
    "Class for easy visualization of point clouds and geometric primitives"
    
    def __init__(self, parent=None):
        "Create a wiget with a VTK Visualizer Control in it"
        self.pointObjects = []
        self.renderer = vtk.vtkRenderer()
        self.renderWidget = RenderWidget(self.renderer, parent)
        self.renderWindow = self.renderWidget.renderWindow
        self.widget = self.renderWidget.widget
        
    def __del__(self):
        del self.renderWidget
                        
    def AddPointCloudActor(self, pc):
        """Add a point cloud from a given NumPy array
        
        The NumPy array should have dimension Nxd where d >= 3
        
        If d>3, the points will be colored according to the last column
        in the supplied array (values should be between 0 and 1, where 
        0 is black and 1 is white)
        """        
        obj = VTKObject()
        obj.CreateFromArray(pc)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddColoredPointCloudActor(self, pc):    
        """Add a point cloud with colors from a given NumPy array
        
        The NumPy array should have dimension Nx6 where the first three
        dimensions correspond to X, Y and Z and the last three dimensions
        correspond to R, G and B values (between 0 and 255)        
        """        
        
        obj = VTKObject()
        obj.CreateFromArray(pc[:,:3])
        obj.AddColors(pc[:,3:6].astype(np.uint8))
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddShadedPointsActor(self, pc):    
        """Add a point cloud with shaded points based on supplied normal vectors
        
        The NumPy array should have dimension Nx6 where the first three
        dimensions correspond to x, y and z and the last three dimensions
        correspond to surface normals (nx, ny, nz)
        """        
        
        obj = VTKObject()
        obj.CreateFromArray(pc[:,0:3])
        obj.AddNormals(pc[:,3:6])
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddPolyDataMeshActor(self,pd):
        "Add a supplied vtkPolyData object to the visualizer"
        obj = VTKObject()
        obj.CreateMeshFromPolyData(pd)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddSTLActor(self,filename):
        "Load a mesh from an STL file and add it to the visualizer" 
        obj = VTKObject()
        obj.CreateFromSTL(filename)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddPLYActor(self,filename):
        "Load a mesh from a PLY file and add it to the visualizer"         
        obj = VTKObject()
        obj.CreateFromPLY(filename)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
        
    def AddNormalsActor(self,pc, scale):
        """Add a set of surface normals to the visualizer
        
        The input is a NumPy array with dimension Nx6 with (x,y,z) and
        (nx,ny,nz) values for the points and associated surface normals
        
        The normals will be scaled according to given scale factor"""
                
        obj = VTKObject()
        obj.CreateFromArray(pc[:,0:3])
        obj.AddNormals(pc[:,3:6])
        obj.SetupPipelineHedgeHog(scale)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddHedgeHogActor(self,pc, scale):        
        """Add shaded points with surface normals to the visualizer
        
        The input is a NumPy array with dimension Nx6 with (x,y,z) and
        (nx,ny,nz) values for the points and associated surface normals
        
        The normals will be scaled according to given scale factor"""
        # Add the points
        self.AddShadedPointsActor(pc)    
        actor = self.GetLastActor()
        actor.GetProperty().SetColor(1,1,1)
        actor.GetProperty().SetPointSize(5.0)
        actor.GetProperty().SetInterpolation(True)
        # Add the normals
        self.AddNormalsActor(pc, scale)        
        
    def AddHedgeHogActorWithScalars(self,pc,scale):   
        """Add shaded points with surface normals and scalars to the visualizer
        
        The input is a NumPy array with dimension Nx7 with (x,y,z),
        (nx,ny,nz) and scalar (the last dimension contains the scalars)
        
        The normals will be scaled according to given scale factor"""
        # Add the points
        self.AddPointCloudActor(pc[:,[0,1,2,-1]])    
        actor = self.GetLastActor()
        actor.GetProperty().SetColor(1,1,1)
        actor.GetProperty().SetPointSize(5.0)
        # Add the normals
        self.AddNormalsActor(pc,scale)  
        
    def AddAxesActor(self,length):
        "Add coordinate system axes with specified length"
        obj = VTKObject()
        obj.CreateAxes(length)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
    
    def AddActor(self,actor):
        "Add a supplied vtkActor object to the visualizer"
        obj = VTKObject()
        obj.CreateFromActor(actor)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
                        
    def ResetCamera(self):
        "Reset the camera to fit contents"
        self.renderer.ResetCamera()
        
    def Render(self):
        "Render all objects"
        self.renderWindow.Render()

    def exec_(self):
        "Run event loop"
        return self.renderWidget.exec_()

    def GetNumberOfActors(self):
        "Get the number of actors added to visualizer"
        return len(self.pointObjects)
        
    def GetLastActorID(self):
        "Get the ID of the last actor added to the visualizer"        
        return len(self.pointObjects)-1
        
    def RemoveAllActors(self):
        "Remove all actors from the visualizer"
        self.pointObjects = []
        self.renderer.RemoveAllViewProps()
        
    def RemoveLastActor(self):
        "Remove the last added actor from the visualizer"
        if len(self.pointObjects) > 0:
            idx = self.GetLastActorID()
            obj = self.pointObjects[idx]
            self.renderer.RemoveActor(obj.GetActor())
            del self.pointObjects[idx]
    
    def GetLastActor(self):
        "Get the actor last added to the visualizer"
        return self.pointObjects[self.GetLastActorID()].GetActor()
        
    def GetActor(self, idx):
        "Get the actor add the specified location"
        return self.pointObjects[idx].GetActor()
        
    def SetActorColor(self, idx, rgb):        
        """Set the color of the specified actor index
        
        rgb is a tuple of (r,g,b) values in the range 0..1
        """
        self.GetActor(idx).GetProperty().SetColor(rgb[0],rgb[1],rgb[2])

    def SetActorOpacity(self, idx, opacity):        
        "Set the opacity (0..1) of the specified actor index"       
        self.GetActor(idx).GetProperty().SetOpacity(opacity)
        
    def SetActorTransform(self, idx, T):
        """Set the pose (location and orientation) of the specified actor index
        
        T is a 4x4 NumPy array containing rotation matrix and translation vector"""
        
        self.GetActor(idx).SetUserTransform(self._array2vtkTransform(T))
        
    def SetActorScale(self, idx, scale):
        "Set the scale of supplied actor index (tuple of 3 values)"
        self.GetActor(idx).SetScale(*scale)
        
    def SetActorVisibility(self, idx,visibility):
        "Toggle visibility of the specified actor index on or of"
        if visibility: 
            self.GetActor(idx).VisibilityOn()
        else:
            self.GetActor(idx).VisibilityOff()
            
    def SetWindowBackground(self,r,g,b):
        "Set the background color of the visualizer to given R, G and B color"
        self.renderer.SetBackground(r,g,b)
        
    def ScreenShot(self,filename):
        "Create a screenshot of the visualizer in BMP format"
        win2img = vtk.vtkWindowToImageFilter()
        win2img.SetInput(self.renderWindow)
        win2img.Update()
        bmpWriter = vtk.vtkBMPWriter()
        bmpWriter.SetInput(win2img.GetOutput())
        bmpWriter.SetFileName(filename)
        bmpWriter.Write()
        
    def Close(self):
        "Close the visualization widget"
        self.widget.close()
        
    # Primitives
    def AddSphere(self, origin, r):
        "Add a sphere with given origin (x,y,z) and radius r"        
        obj = VTKObject()
        obj.CreateSphere(origin, r)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddCylinder(self, origin, r, h):
        "Add a cylinder with given origin (x,y,z), radius r and height h"               
        obj = VTKObject()
        obj.CreateCylinder(origin, r, h)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
                
    def AddPlane(self,normal=None, origin=None):
        """Add a plane (optionally with a given normal vector and origin)
        
        Note: SetActorScale can be used to scale the extent of the plane"""
        obj = VTKObject()
        obj.CreatePlane(normal, origin)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())

    def AddBox(self,bounds):
        "Add a box witih the given bounds=[xmin,xmax,ymin,ymax,zmin,zmax]"        
        obj = VTKObject()
        obj.CreateBox(bounds)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
        
    def AddLine(self,p1,p2):
        "Add a 3D line from p1=[x1,y1,z1] to p2=[x2,y2,z2]"
        obj = VTKObject()
        obj.CreateLine(p1,p2)
        self.pointObjects.append(obj)
        self.renderer.AddActor(obj.GetActor())
        
    # Helper functions    
    def _array2vtkTransform(self,arr):
        T = vtk.vtkTransform()
        matrix = vtk.vtkMatrix4x4()
        for i in range(0,4):
            for j in range(0,4):
                matrix.SetElement(i, j, arr[i,j])
        T.SetMatrix(matrix)
        return T
                                                            
if __name__ == '__main__':

    import sys
    from PyQt5.QtWidgets import *
        
    vtkControl = VTKVisualizerControl()
    
    # Add a red sphere
    vtkControl.AddSphere((0,0,0), 5.0)
    nID = vtkControl.GetLastActorID()
    vtkControl.SetActorColor(nID, (1,0,0))   
    
    # Add a blue, partly transparent box
    vtkControl.AddBox([-10,10,-10,10,-10,10])
    nID = vtkControl.GetLastActorID()
    vtkControl.SetActorColor(nID, (0.8,0.8,1.0))   
    vtkControl.SetActorOpacity(nID, 0.5)
    
        
    app = QApplication.instance()
    
    if app is None:
        app = QApplication(sys.argv)

    app.exec_()            


vtk_visualizer/plot3d.py
# -*- coding: utf-8 -*-
"""
Helper functions for easy visualization of point clouds

.. Author: Øystein Skotheim, SINTEF ICT <oystein.skotheim@sintef.no>
   Date:   Thu Sep 12 15:50:40 2013
"""

import vtk_visualizer.visualizercontrol as vis

g_vtk_control = None
g_hold = False

__colors = {
    'r': (1, 0, 0),
    'g': (0, 1, 0),
    'b': (0, 0, 1),
    'c': (0, 1, 1),
    'm': (1, 0, 1),
    'y': (1, 1, 0),
    'k': (0, 0, 0),
    'w': (1, 1, 1)
}


def _char2color(c):
    "Convert one-letter code to an RGB color"
    global __colors
    return __colors[c]


def get_vtk_control(block):
    "Get the vtk control currently used by the plot3d functions"
    global g_vtk_control

    if block or g_vtk_control is None:
        g_vtk_control = vis.VTKVisualizerControl()

    return g_vtk_control


def toggle_hold(flag=True):
    "Toggle whether new objects will replace previous objects in the visualizer"
    global g_hold
    g_hold = flag


def is_hold_enabled():
    "Returns whether hold is enabled"
    global g_hold
    return g_hold


def plotxyz(pts, color='g', hold=False, block=False):
    """Plot a supplied point cloud (NumPy array of Nxd values where d>=3)

    An optional color may be given as a single character (rgbcmykw)
    An optional hold flag can be enabled to keep previously visualized data

    The supplied array may have an additional column with scalars, which
    will be used to color the points (0=black, 1=white)
    """
    vtkControl = get_vtk_control(block)
    if not (hold or is_hold_enabled()):
        vtkControl.RemoveAllActors()
    vtkControl.AddPointCloudActor(pts)

    if pts.shape[1] <= 3:
        nID = vtkControl.GetLastActorID()
        vtkControl.SetActorColor(nID, _char2color(color))
        vtkControl.Render()
    if block:
        vtkControl.exec_()


def plotxyzrgb(pts, hold=False, block=False):
    "Plot a supplied point cloud w/ color (NumPy array of Nx6 values)"
    vtkControl = get_vtk_control(block)
    if not (hold or is_hold_enabled()):
        vtkControl.RemoveAllActors()
    vtkControl.AddColoredPointCloudActor(pts)
    vtkControl.Render()
    if block:
        vtkControl.exec_()


def plothh(pts, scale=5.0, hold=False, block=False):
    "Plot hedge hog (points w/ normals) from given NumPy array of Nx6 values"
    vtkControl = get_vtk_control(block)
    if not (hold or is_hold_enabled()):
        vtkControl.RemoveAllActors()
    vtkControl.AddHedgeHogActor(pts, scale)
    vtkControl.Render()
    if block:
        vtkControl.exec_()


if __name__ == '__main__':

    import numpy as np
    h, w = 256, 256

    [y, x] = np.mgrid[0:h, 0:w].astype(np.float64)
    z = 10 * np.cos(0.1 * np.sqrt((x - w / 2.0)**2 + (y - h / 2.0)**2))

    pc1 = np.zeros((h, w, 3))
    pc1[:, :, 0] = x
    pc1[:, :, 1] = y
    pc1[:, :, 2] = z

    pc2 = pc1.copy()
    pc2[:, :, 2] += 10.0
    rgb = 255 * np.random.rand(h, w, 3)

    pc2 = np.dstack([pc2, rgb])

    pts1 = pc1.reshape(pc1.shape[0] * pc1.shape[1], pc1.shape[2])
    pts2 = pc2.reshape(pc2.shape[0] * pc2.shape[1], pc2.shape[2])

    plotxyz(pts1, 'r')
    plotxyzrgb(pts2, hold=True)
    import time
    time.sleep(4)
